### PR TITLE
Added retrofit module

### PR DESCRIPTION
**Pull request by AlexeyKorshun**
#203 retrofit call adapter factory for reaktive library.

### CODE REVIEW

File: retrofit/src/main/kotlin/com/badoo/reaktive/retrofit/ReaktiveCallAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+import com.badoo.reaktive.observable.Observable
+import com.badoo.reaktive.observable.asCompletable
+import com.badoo.reaktive.observable.firstOrComplete
+import com.badoo.reaktive.observable.firstOrError
+import retrofit2.Call
+import retrofit2.CallAdapter
+import java.lang.reflect.Type
+
+internal class ReaktiveCallAdapter(
```
Comment (arkivanov) [inline retrofit/src/main/kotlin/com/badoo/reaktive/retrofit/ReaktiveCallAdapter.kt]: Can be replaced with just:
```
internal inline fun <T, R> callAdapter(responseType: Type, crossinline adapt: (Call<T>) -> R): Any =
    object : CallAdapter<T, R> {
        override fun responseType(): Type = responseType

        override fun adapt(call: Call<T>): R = adapt(call)
    }  
```

File: retrofit/src/main/kotlin/com/badoo/reaktive/retrofit/ReaktiveCallAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        private val responseType: Type,
+        private val isBody: Boolean,
```
Comment (arkivanov) [inline retrofit/src/main/kotlin/com/badoo/reaktive/retrofit/ReaktiveCallAdapter.kt]: This set of flags smell, it is possible to specify all of them or any combination. Enum?

File: retrofit/src/main/kotlin/com/badoo/reaktive/retrofit/ReaktiveCallAdapterFactory.kt [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+                check(observableType is ParameterizedType) { "Response must be parameterized as Response<Foo> or Response<? extends Foo>" }
+                responseType = getParameterUpperBound(0, observableType)
+            }
+            else -> {
+                responseType = observableType
+                isBody = true
+            }
+        }
+
+        return ReaktiveCallAdapter(responseType, isBody, isSingle, isMaybe, false)
```
Comment (arkivanov) [inline retrofit/src/main/kotlin/com/badoo/reaktive/retrofit/ReaktiveCallAdapterFactory.kt]: All the code is so Java'ish, here is an idea
```
when {
    rawType == Completable::class.java -> callAdapter(responseType) { it.asSingle().asCompletable() }
    rawType == Single::class.java -> callAdapter(responseType) { it.asSingle() }
    rawType == Maybe::class.java -> callAdapter(responseType) { it.asSingle().asMaybe() }
    rawType == Observable::class.java -> callAdapter(responseType) { it.asSingle().asObservable() }
}
```

File: settings.gradle [hunk 4]
```diff
@@ -15,4 +15,5 @@ if (target.isCommon()) {
     include ':sample-linuxx64-app'
     include ':rxjava2-interop'
     include ':rxjava3-interop'
+    include ':retrofit'
```
Comment (arkivanov) [inline settings.gradle]: I'm not sure that this is a good module name, @CherryPerry what do you think?

**Comment by CherryPerry** [comment]
So, my original opinion was to keep this adapter almost same as in retrofit repository, but I missed, that it is still written in java (even if OkHttp was already migrated to Kotlin).

So I think, this module should be written in more Kotlin-way, not just copy-paste, see @arkivanov proposals.

File: settings.gradle (hunk 4 above)
Comment (CherryPerry) [reply settings.gradle]: Yes, I think `retrofit-adapter` will be much better.

File: retrofit/src/test/kotlin/com/badoo/reaktive/retrofit/StringConverterFactory.kt [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import retrofit2.Retrofit
+import java.lang.reflect.Type
+
+internal class StringConverterFactory : Converter.Factory() {
+
+    override fun responseBodyConverter(
+            type: Type,
+            annotations: Array<Annotation>,
+            retrofit: Retrofit
+    ): Converter<ResponseBody, String> {
```
Comment (CherryPerry) [inline retrofit/src/test/kotlin/com/badoo/reaktive/retrofit/StringConverterFactory.kt]: Single return function.
Convert to 
```kotlin
): Converter<ResponseBody, String> =
    Converter { responseBody -> responseBody.string() }
```

File: retrofit/src/main/kotlin/com/badoo/reaktive/retrofit/CallExecuteObservable.kt [hunk 6]
```diff
@@ -0,0 +1,48 @@
+package com.badoo.reaktive.retrofit
+
+import com.badoo.reaktive.disposable.Disposable
+import com.badoo.reaktive.observable.Observable
+import com.badoo.reaktive.observable.ObservableObserver
+import retrofit2.Call
+import retrofit2.Response
+
+internal class CallExecuteObservable<T>(private val originalCall: Call<T>) : Observable<Response<T>> {
```
[the remark this answers is not in the document]
Comment (CherryPerry) [reply retrofit/src/main/kotlin/com/badoo/reaktive/retrofit/CallExecuteObservable.kt]: I don't remember implementation details, but is not `Call` is synchronous?
Looks like you need to implement [CallEnqueueObservable](https://github.com/square/retrofit/blob/master/retrofit-adapters/rxjava2/src/main/java/retrofit2/adapter/rxjava2/CallEnqueueObservable.java) and not [CallExecuteObservable](https://github.com/square/retrofit/blob/master/retrofit-adapters/rxjava2/src/main/java/retrofit2/adapter/rxjava2/CallExecuteObservable.java). You can add scheduler to factory with `ioScheduler` as default, but I think it is better to reuse thread pool of `OkHttpClient`.
Also you don't need to implement `Callback<Response>` as separate class, use solution provided above.

File: retrofit/src/main/kotlin/com/badoo/reaktive/retrofit/BodyObservable.kt [hunk 7]
```diff
@@ -0,0 +1,42 @@
+package com.badoo.reaktive.retrofit
+
+import com.badoo.reaktive.disposable.Disposable
+import com.badoo.reaktive.observable.Observable
+import com.badoo.reaktive.observable.ObservableObserver
+import retrofit2.HttpException
+import retrofit2.Response
+
+internal class BodyObservable<T>(private val upstream: Observable<Response<T>>) : Observable<T?> {
```
[the remark this answers is not in the document]
Comment (CherryPerry) [reply retrofit/src/main/kotlin/com/badoo/reaktive/retrofit/BodyObservable.kt]: I think that `Single` will be better too. And then convert it to `Observable`, if someone for some strange reason want to use `Observable`.

File: retrofit/build.gradle [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+        jvm(MavenPublication) {
+            from components.java
+        }
+    }
+}
+
+dependencies {
+    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
+    implementation project(':reaktive')
+    implementation project(':reaktive-testing')
```
Comment (CherryPerry) [inline retrofit/build.gradle]: `testImplementation`? What is the reason to bundle testing utils with main package?

File: retrofit/src/test/kotlin/com/badoo/reaktive/retrofit/StringConverterFactory.kt (hunk 5 above)
Comment (CherryPerry) [reply retrofit/src/test/kotlin/com/badoo/reaktive/retrofit/StringConverterFactory.kt]: Also, isn't it easier just to add dependency on `converter-scalars`?

**Comment by AlexeyKorshun** [comment]
@arkivanov @CherryPerry sorry i had not enought time, i am fixing the bugs in this week.

**Comment by CherryPerry** [comment]
Don't worry, take your time.

File: retrofit/src/main/kotlin/com/badoo/reaktive/retrofit/CallExecuteObservable.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (AlexeyKorshun) [reply retrofit/src/main/kotlin/com/badoo/reaktive/retrofit/CallExecuteObservable.kt]: @CherryPerry yes `Call` is synchronous. Why do we need to use Callback?

File: retrofit/src/test/kotlin/com/badoo/reaktive/retrofit/StringConverterFactory.kt (hunk 5 above)
Comment (AlexeyKorshun) [reply retrofit/src/test/kotlin/com/badoo/reaktive/retrofit/StringConverterFactory.kt]: I didn't known about this library. Thx for information.

File: retrofit/src/main/kotlin/com/badoo/reaktive/retrofit/CallExecuteObservable.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (CherryPerry) [reply retrofit/src/main/kotlin/com/badoo/reaktive/retrofit/CallExecuteObservable.kt]: So, original adapter [factory](https://github.com/square/retrofit/blob/master/retrofit-adapters/rxjava2/src/main/java/retrofit2/adapter/rxjava2/RxJava2CallAdapterFactory.java) has 3 methods: `create()`, `createAsync()` and `createWithScheduler()`. I am fine with implementing only one of them, but not `create()`. It is absolutely inconvenient API that makes you set scheduler for each call! `createAsync()` uses OkHttp's tread pool via using Callback API. `createWithScheduler()` uses Call API, but also set valid scheduler via `subscribeOn()`.

So, the best option if you implement all of them.
Or you can just implement `create()` and `createWithScheduler()`.

[the remark this answers is not in the document]
Comment (AlexeyKorshun) [reply retrofit/src/main/kotlin/com/badoo/reaktive/retrofit/CallExecuteObservable.kt]: Hmm. Ok. I know about all of them. But I think `create()` is more flexible way. I'll add `createWithSchedule()` method in this PR and will implement `createAsync()` in separate. @CherryPerry @arkivanov Will that be ok?

File: retrofit-adapter/src/main/kotlin/com/badoo/reaktive/retrofit/ReaktiveCallAdapterFactory.kt [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+            Completable::class.java -> true
+            else -> {
+                check(this is ParameterizedType) { "Return type must be parameterized" }
+                getRawType(getParameterUpperBound(0, this)) != Response::class.java
+            }
+        }
+    }
+}
+
+private fun <T> Single<Response<T>>.transformToBody(transform: Boolean): Single<*> {
```
Comment (CherryPerry) [inline retrofit-adapter/src/main/kotlin/com/badoo/reaktive/retrofit/ReaktiveCallAdapterFactory.kt]: It is single return function.
```
... Single<*> = 
    if (transform) ... then ...
```

Also can you put this function inside `ReaktiveCallAdapterFactory`?

**Comment by CherryPerry** [comment]
Add `retrofit-adapter:build` step to Linux build inside `.github/workflows/build.yml` please.

File: retrofit/src/main/kotlin/com/badoo/reaktive/retrofit/CallExecuteObservable.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (CherryPerry) [reply retrofit/src/main/kotlin/com/badoo/reaktive/retrofit/CallExecuteObservable.kt]: Yes, it is fine.

**Comment by AlexeyKorshun** [comment]
@CherryPerry @arkivanov good day. I fixed the comments, but CI reports a problem with failed tests. The tests were successful at the local machine. But when I cleared my cache I got the 'NoClassDefFoundError: com/badoo/reaktive/test/base/TestObserver' error again. Do you have any ideas what might be wrong?

**Comment by colriot** [comment]
Don't you think that this should be a separate thing? Reaktive is about managing Rx streams, not about network or whatever. At the end of the day, you wouldn't want to keep and maintain everything related to Reaktive in the main repo, right? Imagine adapters for Room, Retrofit, SQLDelight, and lots of others living here.

**Comment by arkivanov** [comment]
@CherryPerry good point above, what do you think? E.g. there is LiveData interop for Reaktive here in a separate repo https://github.com/Lamartio/Reaktive-LiveData-Interop

**Comment by CherryPerry** [comment]
@arkivanov Wrote about it in original issue: https://github.com/badoo/Reaktive/issues/203#issuecomment-533624206. But you decided to add it here.

**Comment by arkivanov** [comment]
Ok my fault, sorry about that. @AlexeyKorshun unfortunately we are not going to accept the adapter for the reasons above. Also we don't want this to become unmaintained as we are not using Retrofit. Feel free to publish it in your or any other repository. We would love to see it published and used by community.

**Pull request closed without merging**